File: sdk/python/tempmail_sdk/providers/tempmailpro.py
"""
TempMail Pro 渠道 - https://tempmailpro.us
"""

from typing import Any, Dict, List, Optional
from urllib.parse import quote

from .. import http as tm_http
from ..normalize import normalize_email
from ..types import Email, EmailInfo

CHANNEL = "tempmailpro"
API_BASE = "https://tempmailpro.us/api/v1"
HEADERS = {"Accept": "application/json", "User-Agent": "Mozilla/5.0"}


def _flatten(raw: Dict[str, Any], recipient: str) -> Dict[str, Any]:
    return {
        "id": raw.get("id") or raw.get("message_id"),
        "from": raw.get("from_address") or raw.get("from_name"),
        "to": recipient,
        "subject": raw.get("subject"),
        "text": raw.get("body_text"),
        "html": raw.get("body_html"),
        "date": raw.get("received_at"),
        "attachments": raw.get("attachments"),
    }


def generate_email() -> EmailInfo:
    resp = tm_http.post(
        f"{API_BASE}/mailbox/create",
        headers={**HEADERS, "Content-Type": "application/json"},
        json={},
        timeout=15,
    )
    resp.raise_for_status()
    data = resp.json()
    if not isinstance(data, dict):
        raise ValueError("tempmailpro: invalid mailbox response")
    box = data.get("data") or {}
    if not isinstance(box, dict):
        raise ValueError("tempmailpro: invalid mailbox response")
    email = str(box.get("address") or "").strip()
    token = str(box.get("token") or "").strip()
    if not email or "@" not in email or not token:
        raise ValueError("tempmailpro: invalid mailbox response")

    return EmailInfo(
        channel=CHANNEL,
        email=email,
        _token=token,
        expires_at=box.get("expires_at"),
        created_at=box.get("created_at"),
    )


def _fetch_detail(token: str, message_id: str) -> Optional[Dict[str, Any]]:
    resp = tm_http.get(
        f"{API_BASE}/mailbox/{quote(token, safe='')}/emails/{quote(message_id, safe='')}",
        headers=HEADERS,
        timeout=15,
    )
    if not resp.ok:
        return None
    data = resp.json()
    if not isinstance(data, dict):
        return None
    detail = data.get("data")
    return detail if isinstance(detail, dict) else None


def get_emails(token: str, email: str) -> List[Email]:
    mailbox_token = (token or "").strip()
    address = (email or "").strip()
    if not mailbox_token:
        raise ValueError("tempmailpro: empty token")
    if not address:
        raise ValueError("tempmailpro: empty email")

    resp = tm_http.get(
        f"{API_BASE}/mailbox/{quote(mailbox_token, safe='')}/emails",
        headers=HEADERS,
        timeout=15,
    )
    resp.raise_for_status()
    data = resp.json()
    if not isinstance(data, dict):
        return []
    rows = data.get("data")
    if not isinstance(rows, list):
        return []

    emails: List[Email] = []
    for item in rows:
        if not isinstance(item, dict):
            continue
        message_id = str(item.get("id") or "").strip()
        detail = _fetch_detail(mailbox_token, message_id) if message_id else None
        emails.append(normalize_email(_flatten(detail or item, address), address))
    return emails

File: sdk/ruby/lib/tempmail_sdk/providers/tempmailpro.rb
# frozen_string_literal: true

require "uri"

module TempmailSdk
  module Providers
    # TempMail Pro 渠道 — https://tempmailpro.us
    # POST /api/v1/mailbox/create 创建邮箱，GET /api/v1/mailbox/{token}/emails 获取邮件
    module Tempmailpro
      CHANNEL = "tempmailpro"
      API_BASE = "https://tempmailpro.us/api/v1"
      HEADERS = { "Accept" => "application/json", "User-Agent" => "Mozilla/5.0" }.freeze

      module_function

      def flatten(raw, recipient)
        {
          "id" => raw["id"] || raw["message_id"],
          "from" => raw["from_address"] || raw["from_name"],
          "to" => recipient,
          "subject" => raw["subject"],
          "text" => raw["body_text"],
          "html" => raw["body_html"],
          "date" => raw["received_at"],
          "attachments" => raw["attachments"]
        }
      end

      # @return [EmailInfo]
      def generate_email
        resp = Http.post("#{API_BASE}/mailbox/create",
                         headers: HEADERS.merge("Content-Type" => "application/json"),
                         json: {}, timeout: 15)
        resp.raise_for_status
        data = resp.json
        raise "tempmailpro: 创建邮箱响应无效" unless data.is_a?(Hash)

        box = data["data"] || {}
        raise "tempmailpro: 创建邮箱响应无效" unless box.is_a?(Hash)

        email = (box["address"] || "").strip
        token = (box["token"] || "").strip
        raise "tempmailpro: 创建邮箱响应无效" if email.empty? || !email.include?("@") || token.empty?

        EmailInfo.new(channel: CHANNEL, email: email, token: token,
                      expires_at: box["expires_at"], created_at: box["created_at"])
      end

      # @param token [String]
      # @param email [String]
      # @return [Array<Email>]
      def get_emails(token, email)
        mailbox_token = (token || "").strip
        address = (email || "").strip
        raise "tempmailpro: token 为空" if mailbox_token.empty?
        raise "tempmailpro: 邮箱地址为空" if address.empty?

        resp = Http.get("#{API_BASE}/mailbox/#{URI.encode_www_form_component(mailbox_token)}/emails",
                        headers: HEADERS, timeout: 15)
        resp.raise_for_status
        data = resp.json
        return [] unless data.is_a?(Hash)

        rows = data["data"]
        return [] unless rows.is_a?(Array)

        rows.filter_map do |item|
          next unless item.is_a?(Hash)

          message_id = (item["id"] || "").to_s.strip
          detail = nil
          unless message_id.empty?
            begin
              dr = Http.get("#{API_BASE}/mailbox/#{URI.encode_www_form_component(mailbox_token)}/emails/#{URI.encode_www_form_component(message_id)}",
                            headers: HEADERS, timeout: 15)
              if dr.ok?
                dd = dr.json
                detail = dd["data"] if dd.is_a?(Hash) && dd["data"].is_a?(Hash)
              end
            rescue StandardError
              # 忽略
            end
          end
          Normalize.normalize_email(flatten(detail || item, address), address)
        end
      end
    end
  end
end
